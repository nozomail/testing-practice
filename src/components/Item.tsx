import React, { useState } from "react";
import { ListItem, Box } from "@material-ui/core";
import LocalBar from "@material-ui/icons/LocalBarRounded";
import { DataProps } from "../types";

function Item({ idDrink, strAlcoholic, strDrink, strInstructions, strIngredients }: DataProps) {
  const [isOpen, setisOpen] = useState(false);

  return (
    <ListItem component="li" button key={idDrink} onClick={() => setisOpen(!isOpen)}>
      <div>
        <strong>{strDrink}</strong>
        {strAlcoholic === "Alcoholic" && <LocalBar color="secondary" style={{ fontSize: "16px", marginLeft: "0.25em" }} data-testid="alcoholic" />}
        {isOpen && (
          <Box pt={2}>
            <Box mb={1}>Ingredients:</Box>
            <ul>
              {strIngredients.map((ingredient, index) => (
                <li key={ingredient + index}>{ingredient}</li>
              ))}
            </ul>
            <p>{strInstructions}</p>
          </Box>
        )}
      </div>
    </ListItem>
  );
}

export default Item;
