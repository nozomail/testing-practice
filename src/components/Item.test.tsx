import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Item from "./Item";

describe("Item", () => {
  const item1 = {
    idDrink: "12564",
    strAlcoholic: "Non alcoholic",
    strDrink: "Apple Karate",
    strInstructions: "Place all ingredients in the blender jar - cover and whiz on medium speed until well blended. Pour in one tall, 2 medium or 3 small glasses and drink up.",
    strIngredients: ["Carrot"],
  };
  const item2 = {
    idDrink: "11046",
    strAlcoholic: "Alcoholic",
    strDrink: "Applecar",
    strInstructions: "Shake all ingredients with ice, strain into a cocktail glass, and serve.",
    strIngredients: ["Applejack", "Triple sec", "Lemon juice"],
  };

  test("matches snapshot", () => {
    const { asFragment } = render(<Item {...item1} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("shows item's name without alcoholic icon", () => {
    render(<Item {...item1} />);

    expect(screen.getByText(/Apple Karate/i)).toBeInTheDocument();
    expect(screen.queryByTestId("alcoholic")).not.toBeInTheDocument();
  });

  test("shows item's name with alcoholic icon", () => {
    render(<Item {...item2} />);

    expect(screen.getByText(/Applecar/i)).toBeInTheDocument();
    expect(screen.queryByTestId("alcoholic")).toBeInTheDocument();
  });

  test("does not show neither ingredients nor instructions", () => {
    render(<Item {...item1} />);

    expect(screen.queryByText(/Ingredients:/i)).not.toBeInTheDocument();
    expect(screen.queryByText(/Carrot/i)).not.toBeInTheDocument();
    expect(screen.queryByText(/Place all ingredients/i)).not.toBeInTheDocument();
  });

  test("shows ingredients and instructions when clicked", () => {
    render(<Item {...item1} />);

    userEvent.click(screen.getByRole("button"));

    expect(screen.getByText(/Ingredients:/i)).toBeInTheDocument();
    expect(screen.getByText(/Carrot/i)).toBeInTheDocument();
    expect(screen.getByText(/Place all ingredients/i)).toBeInTheDocument();
  });

  test("hides ingredients and instructions when clicked again", () => {
    render(<Item {...item1} />);

    const target = screen.getByRole("button");
    userEvent.click(target);
    userEvent.click(target);

    expect(screen.queryByText(/Ingredients:/i)).not.toBeInTheDocument();
    expect(screen.queryByText(/Carrot/i)).not.toBeInTheDocument();
    expect(screen.queryByText(/Place all ingredients/i)).not.toBeInTheDocument();
  });
});
