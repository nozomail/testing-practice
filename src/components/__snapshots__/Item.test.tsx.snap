// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item matches snapshot 1`] = `
<DocumentFragment>
  <li
    aria-disabled="false"
    class="MuiButtonBase-root MuiListItem-root MuiListItem-gutters MuiListItem-button"
    role="button"
    tabindex="0"
  >
    <div>
      <strong>
        Apple Karate
      </strong>
    </div>
    <span
      class="MuiTouchRipple-root"
    />
  </li>
</DocumentFragment>
`;
